File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/debug/ProxyGuidanceTest.kt
package com.tyhp.lang.debug

import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ProxyGuidanceTest {
    @Test
    fun `Tyhp remote debug configs mention tyhp`() {
        assertTrue(isTyhpPhpRemoteDebugConfig(TYHP_PHP_DEBUG_CONFIG_NAME, PHP_REMOTE_DEBUG_TYPE_ID))
        assertFalse(isTyhpPhpRemoteDebugConfig("Listen for Xdebug", PHP_REMOTE_DEBUG_TYPE_ID))
        assertFalse(isTyhpPhpRemoteDebugConfig("Listen for Tyhp", "NodeJSConfigurationType"))
    }

    @Test
    fun `missing PHP Remote Debug guidance names PhpStorm XDebug`() {
        assertContains(phpRemoteDebugMissingGuidance(), "PHP Remote Debug")
        assertContains(phpRemoteDebugMissingGuidance(), XDEBUG_PROXY_DOCS_URL)
    }

    @Test
    fun `proxy-down guidance names the IDE port and start command`() {
        val text = proxyDownGuidance(9003)
        assertContains(text, "9003")
        assertContains(text, "Start XDebug Proxy")
    }

    @Test
    fun `sourcemap guidance when generateSourcemap is off`() {
        val text = sourcemapGuidance(SourcemapGuidanceOptions(generateSourcemap = false))
        assertNotNull(text)
        assertContains(text, "generateSourcemap")
        assertContains(text, SOURCEMAP_DOCS_URL)
    }

    @Test
    fun `sourcemap guidance when maps are missing after a build flag is on`() {
        val text = sourcemapGuidance(
            SourcemapGuidanceOptions(
                generateSourcemap = true,
                mapCount = 0,
                sourceMapDir = "./build/",
            ),
        )
        assertNotNull(text)
        assertContains(text, ".php.map")
        assertContains(text, "build")
    }

    @Test
    fun `no sourcemap guidance when maps are present`() {
        assertNull(sourcemapGuidance(SourcemapGuidanceOptions(generateSourcemap = true, mapCount = 3)))
    }
}

File: tyhp-lang/vscode/src/debug/proxyGuidance.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
    PHP_DEBUG_EXTENSION_ID,
    TYHP_PHP_DEBUG_CONFIG_NAME,
    isTyhpPhpDebugConfig,
    phpDebugMissingGuidance,
    proxyDownGuidance,
    sourcemapGuidance,
} from "./proxyGuidance";

test("Tyhp launch configs are php configs whose name mentions Tyhp", () => {
    assert.equal(
        isTyhpPhpDebugConfig({ type: "php", name: TYHP_PHP_DEBUG_CONFIG_NAME }),
        true
    );
    assert.equal(isTyhpPhpDebugConfig({ type: "php", name: "Listen for Xdebug" }), false);
    assert.equal(isTyhpPhpDebugConfig({ type: "node", name: "Listen for Tyhp" }), false);
});

test("missing PHP Debug guidance names the extension id", () => {
    assert.match(phpDebugMissingGuidance(), new RegExp(PHP_DEBUG_EXTENSION_ID));
});

test("proxy-down guidance names the IDE port and start command", () => {
    const text = proxyDownGuidance(9003);
    assert.match(text, /9003/);
    assert.match(text, /Start XDebug Proxy/);
});

test("sourcemap guidance when generateSourcemap is off", () => {
    const text = sourcemapGuidance({ generateSourcemap: false });
    assert.ok(text);
    assert.match(text ?? "", /generateSourcemap/);
});

test("sourcemap guidance when maps are missing after a build flag is on", () => {
    const text = sourcemapGuidance({
        generateSourcemap: true,
        mapCount: 0,
        sourceMapDir: "./build/",
    });
    assert.ok(text);
    assert.match(text ?? "", /\.php\.map/);
    assert.match(text ?? "", /build/);
});

test("no sourcemap guidance when maps are present", () => {
    assert.equal(
        sourcemapGuidance({ generateSourcemap: true, mapCount: 3 }),
        undefined
    );
});
